### ISSUE TITLE
NotificationQueue

### ISSUE THREAD

**Issue by micheleselea** [pocoproject/poco#827]
Hi all, I don't know if this is an issue or it's the correct way to work, I only report this to save some deadlock.
I found that if a thread is waiting for notification and makes some slow tasks so

do{
Poco::Notification::Ptr nf=m_nq.waitDequeueNotification();
if(nf.isNull())
break;
if(bWantToStop)
break;
slow task
}while(1);

if  from another thread stop the execution with
m_nq.wakeUpAll()
th.join();

the thread stops, no problem: but if during the slow task process someone push a new notification with enqueueNotification...thread never stops

**Comment by aleks-f** [comment]
I have experienced similar problem.

It may not be same as your case, but if it is, then it doesn't happen because you have new notifications enqueued  - in such a case ```waitDequeueNotification()``` would return and the ```bWantToStop``` flag (I assume you are setting it before ```wakeUpAll()``` call) should terminate the thread loop. However, if ```waitDequeueNotification()``` is called immediately after ```wakeUpAll()``` and there are no new notifications, it will block forever waiting while you are trying to join the blocked thread. I have solved my problem by calling ```waitDequeueNotification(100)``` to make sure it will always return, even if a bit delayed.

@obiltschnig : are we missing something here?
